kolibri/core/device/utils.py

```diff
@@ -141,3 +111,219 @@ def set_app_key_on_response(response):
+    if language_id is not None:
+        settings_to_set["language_id"] = language_id
+    if facility is not None:
+        settings_to_set["default_facility"] = facility
+
+    return settings_to_set
+
+
+def create_facility(facility_name=None, preset=None):
+    from kolibri.core.auth.models import Facility
+
+    facility = Facility.objects.create(name=facility_name)
+    logger.info("Facility with name '{name}' created.".format(name=facility.name))
+
+    # Only set preset data if we have created the facility, otherwise leave previous data intact
+    if preset:
+        dataset_data = mappings[preset]
+        facility.dataset.preset = preset
+        for key, value in dataset_data.items():
+            setattr(facility.dataset, key, value)
+        facility.dataset.save()
+        logger.info("Facility preset changed to {preset}.".format(preset=preset))
+    return facility
+
+
+def setup_device_and_facility(
+    facility,
+    facility_name,
+    preset,
+    facility_settings,
+    device_settings,
+    username,
+    password,
+):
+    from kolibri.core.auth.models import FacilityUser
+
+    with transaction.atomic():
+        if facility is None and facility_name is not None:
+            facility = create_facility(
+                facility_name=facility_name,
+                preset=preset,
+            )
+
+            if facility_settings:
+                for key, value in facility_settings.items():
+                    setattr(facility.dataset, key, value)
+                facility.dataset.save()
+                logger.info(
+                    "Facility settings updated with {}".format(facility_settings)
+                )
+
+        provision_device(**device_settings)
+        logger.info("Device settings updated with {}".format(device_settings))
+
+        if username and password and facility:
+            try:
+                FacilityUser.objects.create_superuser(
+                    username, password, facility=facility
+                )
+                logger.info(
+                    "Superuser created with username '{username}' in facility '{facility}'.".format(
+                        username=username, facility=facility
+                    )
+                )
+            except ValidationError:
+                logger.warn(
+                    "An account with username '{username}' already exists in facility '{facility}', not creating user account.".format(
+                        username=username, facility=facility
+                    )
+                )
+
+
+def provision_from_file(file_path):
+    """
+    Expects a JSON file with the following format (example values supplied):
+    {
+        "facility": "My Facility",
+        "preset": "formal",
+        "facility_settings": {
+            "learner_can_edit_username": true,
+            "learner_can_edit_name": true,
+            "learner_can_edit_password": true,
+            "learner_can_sign_up": true,
```

Comment:
In previous work on so-called ["device templates"](https://www.notion.so/learningequality/Data-and-content-architecture-proposal-8e03a6da67394c578d40336b7a7953bf#5a856f31bfcf49e29678a7bb5dec342c), I recall there may have been use-cases where we sometimes wanted these values to be fixed, and other times we wanted them to be treated as defaults with the expectation that users with sufficient permissions could modify them. I don't recall specific scenarios.

Here's the relevant snippet:

> Under the hood, management of certain device properties can be delegated to a facility user with 'sufficient' permission. The template can specify assign one of four device template permission delegation strategies to a particular device property:
> 
> 1. **No delegation:** the property is fixed at the default value.
> 2. **By user role:** the property can be modified by certain types of user (learner, coach, admin). For example, facility admins might have the ability to manage content, but not facilities or device settings.
> 3. **By user permissions:** for example, a particular user might have `can_manage_content=True`. We can add additional user properties like `can_manage_facilities` and `can_manage_device_settings`.
> 4. **By both user role and permissions.**
> 
> Note that delegation can be considered an implementation detail, and would not necessarily be exposed to users in full. It's more likely that we would define fewer, simpler device management options to handle the most common use-cases and expose those options to users.

This whole delegation system might not be exactly what we need, but I could imagine possibly having settings take _objects_ instead of raw values to allow for future extensibility.

For example we could do something like this to distinguish default from permanently fixed values:

```suggestion
            "learner_can_sign_up": { "value": true, "fixed": true }
```

Reply:
I would be OK with keeping it simple for now, but allowing either of these in the future. Seems OK to have the shorthand for the defaults.